rosidl_parser/rosidl_parser/definition.py

```diff
@@ -0,0 +1,343 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+import pathlib
+
+
+BASIC_TYPES = [
+    'short',
+    'unsigned short',
+    'long',
+    'unsigned long',
+    'long long',
+    'unsigned long long',
+    'float',
+    'double',
+    'long double',
+    'char',
+    'wchar',
+    'boolean',
+    'octet',
+    'int8',
+    'uint8',
+    'int16',
+    'uint16',
+    'int32',
+    'uint32',
+    'int64',
+    'uint64',
+]
+
+SERVICE_REQUEST_MESSAGE_SUFFIX = '_Request'
+SERVICE_RESPONSE_MESSAGE_SUFFIX = '_Response'
+
+
+class AbstractType:
+
+    __slots__ = ()
+
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+
+class BaseType(AbstractType):
+
+    __slots__ = ()
+
+
+class BasicType(BaseType):
+
+    __slots__ = ('type', )
+
+    def __init__(self, typename):
+        super().__init__()
+        assert typename in BASIC_TYPES
+        self.type = typename
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.type == other.type
+
+
+class NamedType(BaseType):
+
+    __slots__ = ('name')
+
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.name == other.name
+
+
+class NamespacedType(BaseType):
+
+    __slots__ = ('namespaces', 'name')
+
+    def __init__(self, namespaces, name):
+        super().__init__()
+        self.namespaces = namespaces
```

Comment:
Nit, my first thought was that the plural name meant a type could have multiple fully qualified namespaces, but now I see that it's a list of the different names making up a single namespace. Would you mind adding a comment to this type to describe it?

Reply:
Added docblock in 7647c6b042bd4c0d4135baf513c34948ec82e729.